added toxav_group_bit_rate_set

Dako300 opened the pull request:
I tested this with a forked qTox in a call with myself, and it seems to work fine.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/c-toxcore/464)
<!-- Reviewable:end -->

Dako300 commented:
Turns out this works fine with two people, but three or more experience a lot of audio loss

Dako300 commented:
The group call issue was a client problem. It appears this works fine and also allows for on the fly changing of values.

sudden6 commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[toxav/groupav.c, line 29 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KcSGAUZ5dgDQ38-TD7k:-KcSGAUZ5dgDQ38-TD7l:b-kd4ump) ([raw file](https://github.com/toktok/c-toxcore/blob/ac6e3f8082e02773a16e7173d4047823c518a839/toxav/groupav.c#L29)):*
> ```C
> #include "../toxcore/util.h"
> 
> #include "toxav.h" // messy, for audio_bit_rate_invalid
> ```

I don't think you're supposed to include the api header

---

*[toxav/groupav.c, line 250 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KcSFBRYp84r2UTXKahd:-KcSFBRZoWIS1RTOkgiZ:b-3k53br) ([raw file](https://github.com/toktok/c-toxcore/blob/ac6e3f8082e02773a16e7173d4047823c518a839/toxav/groupav.c#L250)):*
> ```C
> 
>     group_av->audio_bitrate = 16000;
>     
> ```

Whitespace errors

---

*[toxav/groupav.c, line 547 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KcSGFHDwrl5ncaG82tW:-KcSGFHDwrl5ncaG82tX:bpgqee3) ([raw file](https://github.com/toktok/c-toxcore/blob/ac6e3f8082e02773a16e7173d4047823c518a839/toxav/groupav.c#L547)):*
> ```C
> 
>     if(audio_bit_rate_invalid(audio_bit_rate)) {
> 	return -1;
> ```

whitespace errors here too

---

*[toxav/groupav.h, line 21 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KcSGJAsESZrCQPYEjiH:-KcSGJAt2mjQ2RCnvEY1:b-kd4ump) ([raw file](https://github.com/toktok/c-toxcore/blob/ac6e3f8082e02773a16e7173d4047823c518a839/toxav/groupav.h#L21)):*
> ```objc
>  */
> #include "../toxcore/group.h"
> #include "toxav.h" // bit_rate
> ```

I don't think you're supposed to include the api header

---

*[toxav/toxav.h, line 762 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KcSFknfXpH20hhmJYPL:-KcSFkngrmhsHHIbCdFI:bknz70p) ([raw file](https://github.com/toktok/c-toxcore/blob/ac6e3f8082e02773a16e7173d4047823c518a839/toxav/toxav.h#L762)):*
> ```objc
> 
> 
> /**
> ```

don't change the header directly, it's generated from toxav.api.h

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
Min and Max bitrates should be provided with getter functions

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-:-KcSGfNRDe7FFjTC-50B:b-crjsuj)*
<!-- Sent from Reviewable.io -->

Dako300 commented:
Fixed everything pointed out so far.

sudden6 commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[toxav/groupav.c, line 547 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KcSGFHDwrl5ncaG82tW:-KcYY9gz2Yk6uRBOOe6S:bqag9kj) ([raw file](https://github.com/toktok/c-toxcore/blob/ac6e3f8082e02773a16e7173d4047823c518a839/toxav/groupav.c#L547)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

whitespace errors here too
</blockquote></details>

there are still tabs where spaces should be

---

*[toxav/toxav.h, line 762 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KcSFknfXpH20hhmJYPL:-KcYZO176048zYgC9GnT:b-51nom) ([raw file](https://github.com/toktok/c-toxcore/blob/ac6e3f8082e02773a16e7173d4047823c518a839/toxav/toxav.h#L762)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

don't change the header directly, it's generated from toxav.api.h
</blockquote></details>

you have to regenerate the header, see https://github.com/TokTok/c-toxcore/blob/master/docs/apidsl.md for the procedure

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464)*
<!-- Sent from Reviewable.io -->

Dako300 commented:
I fixed both of those errors

sudden6 commented:
Reviewed 1 of 4 files at r1, 3 of 4 files at r2.
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[toxav/groupav.h, line 27 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-Kcconkdt7KxeXPG-IUy:-KcconkerB3K59JsnJYF:b-wyrh4a) ([raw file](https://github.com/toktok/c-toxcore/blob/cc4ea69dc238bac5beb80ab23d5bb1e0a8938ba5/toxav/groupav.h#L27)):*
> ```objc
> #define GROUP_AUDIO_PACKET_ID 192
> 
> #define INVALID_AUDIO_BITRATE(a) (a < 6 || a > 510)
> ```

Having both, this define and the function is dangerous, because they have to be kept in sync, is there a way to define this only once?
Additionally I think there should be getter functions for the min and max bitrates.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464)*
<!-- Sent from Reviewable.io -->

Dako300 commented:
Ill see what i can do

iphydf commented on toxav/groupav.c:
```diff
@@ -525,6 +527,33 @@ static int send_audio_packet(Group_Chats *g_c, int groupnumber, uint8_t *packet,
+{
+    Group_AV *group_av = (Group_AV *)group_get_object(g_c, groupnumber);
+
+    if(!group_av) {
```
Add space before `(` in all `if` statements.

iphydf commented on toxav/groupav.c:
```diff
@@ -525,6 +527,33 @@ static int send_audio_packet(Group_Chats *g_c, int groupnumber, uint8_t *packet,
+        return -1;
+    }
+
+    if(group_av->audio_bitrate/1000 != audio_bit_rate) {
```
Add spaces around operator `/`.

iphydf commented on toxav/toxav.api.h:
```diff
@@ -628,6 +628,18 @@ int toxav_join_av_groupchat(Tox *tox, int32_t friendnumber, const uint8_t *data,
+
+int toxav_group_bit_rate_set(Tox *tox, int groupnumber, int32_t audio_bit_rate);
+
+bool audio_bit_rate_invalid(uint32_t bit_rate);
```
Add documentation. Also, this function should probably test for validity, not for invalidity. Also, it needs a toxav namespace prefix.

Dako300 replied:
The audio_bit_rate_invalid function existed before I forked it, I just made defined it in the header so I can reference it in the groupav.c file. I'm changing the header version to be toxav_audio_bit_rate_valid

iphydf replied:
Ok, let me know when the change is done.

Dako300 replied:
All of the changes were applied, but my fork fell out of date. I'll apply this to the newer version

iphydf commented on toxav/toxav.api.h:
```diff
@@ -628,6 +628,18 @@ int toxav_join_av_groupchat(Tox *tox, int32_t friendnumber, const uint8_t *data,
+ * return -1 on failure.
+ */
+
+int toxav_group_bit_rate_set(Tox *tox, int groupnumber, int32_t audio_bit_rate);
```
How did you decide to make audio_bit_rate a signed int32_t?

Dako300 replied:
I made it signed because I modeled the template for the function call off of toxav_bit_rate_set.

iphydf replied:
I see, toxav_call and toxav_bit_rate_set are inconsistent. Can you make it unsigned? There are no negative bit rates.

Dako300 commented:
I added min and max getters for group audio bit rate (min and max are defined as min and max per the Opus spec), rebased my fork against the latest commit on the main branch, and fixed all problems raised so far that I could see.

nurupo commented:
Just want to not that this PR [doesn't pass the format test](https://travis-ci.org/TokTok/c-toxcore/jobs/208031049#L2450-L2556).

Please [astyle](https://github.com/TokTok/c-toxcore/tree/master/other/astyle) your code. Also make sure not to edit `toxav.h` directly. `toxav.h` must be generated off `toxav.api.h` using [APIDSL](https://github.com/TokTok/apidsl).

sudden6 commented:
Reviewed 2 of 7 files at r3.
Review status: 2 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[toxav/groupav.c, line 555 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-Ken2rNZh1NJDsjW0XbA:-Ken2rN_LGNEebc-0hxx:b-bdq992) ([raw file](https://github.com/toktok/c-toxcore/blob/59eefd73068599dbc885ffc609aa5a7908b6f08e/toxav/groupav.c#L555)):*
> ```C
>         }
>     }
>     
> ```

whitespace error

---

*[toxcore/tox.h, line 183 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-Ken3SNj8FGDFFrMNn-d:-Ken3SNj8FGDFFrMNn-e:b-nyj8p6) ([raw file](https://github.com/toktok/c-toxcore/blob/59eefd73068599dbc885ffc609aa5a7908b6f08e/toxcore/tox.h#L183)):*
> ```objc
>  * changing any functionality or API or ABI.
>  */
> #define TOX_VERSION_PATCH              6
> ```

I don't think the version should be changed in this PR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Please enable the checkbox "Allow edits from maintainers." on the bottom right.

Also, it now conflicts because of the version change. Rebase on master should fix that.

What issue does this solve?

Dako300 commented:
@iphydf This allows for changes in the group audio bitrate. Before this, they were defined as constants inside of the call functions. I checked that allow edits box.

iphydf commented:
@mannol can you review this?

iphydf commented:
Review status: 2 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[toxav/groupav.h, line 57 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KhZ2b_lLUVq25XoVo97:-KhZ2b_lLUVq25XoVo98:b-cchzdy) ([raw file](https://github.com/toktok/c-toxcore/blob/bb3cc2defd04122c11e7ce7cb7f97e368088117c/toxav/groupav.h#L57)):*
> ```objc
> 
> bool group_bit_rate_valid(uint32_t audio_bit_rate);
> uint32_t group_min_bit_rate();
> ```

`(void)` instead of `()`, otherwise this is an old style "unknown argument list".

---

*[toxav/toxav.api.h, line 641 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KcV2cVUTFEeYWE3flUT-r2-641:-KhZ2ynoKTNeoPdkfNy7:b-5qsy8w) ([raw file](https://github.com/toktok/c-toxcore/blob/cc4ea69dc238bac5beb80ab23d5bb1e0a8938ba5/toxav/toxav.api.h#L641)):*
<details><summary><i>Previously, Dako300 wrote…</i></summary><blockquote>

All of the changes were applied, but my fork fell out of date. I'll apply this to the newer version
</blockquote></details>

Add docs?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464)*
<!-- Sent from Reviewable.io -->

mannol commented:
*[toxav/groupav.c, line 551 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KhaUAjNAhzbwvmXnhNc:-KhaUAjNAhzbwvmXnhNd:b-96u1pz) ([raw file](https://github.com/toktok/c-toxcore/blob/bb3cc2defd04122c11e7ce7cb7f97e368088117c/toxav/groupav.c#L551)):*
> ```C
>         group_av->audio_bitrate = audio_bit_rate * 1000;
> 
>         if (recreate_encoder(group_av) == -1) {
> ```

If this fails, the group_av->audio_bitrate will remain set on a new value so it can cause problems in future retries.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464)*
<!-- Sent from Reviewable.io -->

mannol commented:
I've requested a change.

More explanation: The reconfiguration is quite tricky, if it fails you've lost your good encoder without a way of falling back. I propose a use of create_audio_encoder() function as seen in audio.c

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-:-KhaVjSbNZn05GvdGfQU:b-1kuho6)*
<!-- Sent from Reviewable.io -->

mannol commented:
Reviewed 6 of 7 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464)*
<!-- Sent from Reviewable.io -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/TokTok/c-toxcore?pullRequest=464) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/TokTok/c-toxcore?pullRequest=464) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/TokTok/c-toxcore?pullRequest=464) it.</sub>

robinlinden commented:
Reviewed 4 of 7 files at r3, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[toxav/toxav.c, line 1010 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KwS8YVScvF5JKkUbGzM:-KwS8YVScvF5JKkUbGzN:bj5ywmh) ([raw file](https://github.com/toktok/c-toxcore/blob/68a5aa1233ca23fffde15df5665a5edcc46d0236/toxav/toxav.c#L1010)):*
> ```C
>      * Opus supports all bit rates from 6 kbit/s to 510 kbit/s.
>      */
>     return bit_rate < 6 || bit_rate > 510;
> ```

All places in the code where the min and max bit rates are used should use the new functions. These constants should appear in the code exactly once.

---

*[toxav/toxav.c, line 1014 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-KwS8nv5XPbvYLVIi_sV:-KwS8nv5XPbvYLVIi_sW:boibw3z) ([raw file](https://github.com/toktok/c-toxcore/blob/68a5aa1233ca23fffde15df5665a5edcc46d0236/toxav/toxav.c#L1014)):*
> ```C
> bool toxav_audio_bit_rate_valid(uint32_t bit_rate)
> {
>     return !audio_bit_rate_invalid(bit_rate);
> ```

This looks weird. You should probably change the `audio_bit_rate_invalid` function to check if the bit rate's valid instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464)*
<!-- Sent from Reviewable.io -->

iphydf commented:
@mannol @sudden6 have you reviewed this?
@Dako300 can you rebase this on upstream master?

sudden6 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[toxav/groupav.c, line 603 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/464#-L2khnCY4IC8QXLJOX6U:-L2khnCY4IC8QXLJOX6V:b-4fo4pd) ([raw file](https://github.com/toktok/c-toxcore/blob/68a5aa1233ca23fffde15df5665a5edcc46d0236/toxav/groupav.c#L603)):*
> ```C
> }
> 
> uint32_t group_min_bit_rate()
> ```

There should be a comment somewhere that this value is in kbit/s or IMO better to change the code to use bit/s

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/464)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
There are a lot of open comments so I can't yet give lgtm.

iphydf commented:
@Dako300 are you still going to work on this?

iphydf commented:
Closing for inactivity. Please reopen if you're going to work on this.
